crates/qpy/src/circuits.rs

```diff
@@ -0,0 +1,2174 @@
+            .gate
+            .getattr(py, "num_ctrl_qubits")?
+            .extract::<u32>(py),
+        OperationRef::Instruction(py_inst) => py_inst
+            .instruction
+            .getattr(py, "num_ctrl_qubits")?
+            .extract::<u32>(py),
+        _ => Ok(0),
+    }
+}
+
+fn get_ctrl_state(py: Python, op: &PackedOperation, num_ctrl_qubits: u32) -> PyResult<u32> {
+    match op.view() {
+        OperationRef::Gate(py_gate) => py_gate.gate.getattr(py, "ctrl_state")?.extract::<u32>(py),
+        OperationRef::Instruction(py_inst) => py_inst
+            .instruction
+            .getattr(py, "ctrl_state")?
+            .extract::<u32>(py),
+        _ => Ok((1 << num_ctrl_qubits) - 1),
+    }
+}
+
+fn get_instruction_params(
+    py: Python,
+    instruction: &PackedInstruction,
+    qpy_data: &QPYWriteData,
+) -> PyResult<Vec<formats::PackedParam>> {
+    // The instruction params we store are about being able to reconstruct the objects; they don't
+    // necessarily need to match one-to-one to the `params` field.
+    if let OperationRef::Instruction(inst) = instruction.op.view() {
+        if inst
+            .instruction
+            .bind(py)
+            .is_instance(imports::CONTROL_FLOW_SWITCH_CASE_OP.get_bound(py))?
+        {
+            let op = inst.instruction.bind(py);
+            let target = op.getattr("target")?;
+            let cases = op.call_method0("cases_specifier")?;
+            let cases_tuple = imports::BUILTIN_TUPLE.get_bound(py).call1((cases,))?;
+            return Ok(vec![
+                pack_param(py, &target.extract::<Param>()?, qpy_data)?,
+                pack_param(py, &cases_tuple.extract::<Param>()?, qpy_data)?,
+            ]);
+        }
+        if inst
+            .instruction
+            .bind(py)
+            .is_instance(imports::CONTROL_FLOW_BOX_OP.get_bound(py))?
+        {
+            let op = inst.instruction.bind(py);
+            let first_block = op
+                .getattr("blocks")?
+                .try_iter()?
+                .next()
+                .transpose()?
+                .ok_or_else(|| PyValueError::new_err("No blocks in box control flow op"))?;
+            let duration = op.getattr("duration")?;
+            let unit = op.getattr("unit")?;
+            return Ok(vec![
+                pack_param(py, &first_block.extract::<Param>()?, qpy_data)?,
+                pack_param(py, &duration.extract::<Param>()?, qpy_data)?,
+                pack_param(py, &unit.extract::<Param>()?, qpy_data)?,
+            ]);
+        }
+    }
+
+    if let OperationRef::Operation(op) = instruction.op.view() {
+        if op
+            .operation
+            .bind(py)
+            .is_instance(imports::CLIFFORD.get_bound(py))?
+        {
+            let op = op.operation.bind(py);
+            let tableau = op.getattr("tableau")?;
+            return Ok(vec![pack_param(
+                py,
+                &tableau.extract::<Param>()?,
```

Comment:
Same comment 🙂.

I think there are a lot of instance where you're using `Param` to represent things that are not actually `Param` but "parameters" in the sense that they are parameters we need passed to the constructor of the Python op in question.